--- libiscsi-1.9.0/configure.ac~	2013-06-20 21:03:22.583278957 +0200
+++ libiscsi-1.9.0/configure.ac	2013-06-20 21:04:57.467956077 +0200
@@ -12,7 +12,7 @@ AM_CONDITIONAL(LD_ISCSI,
                [expr "$host_os" : linux > /dev/null 2>&1])
 
 if test "$ac_cv_prog_gcc" = yes; then
-   WARN_CFLAGS="-Wall -W -Wshadow -Wstrict-prototypes -Wpointer-arith -Wcast-align -Wwrite-strings -Werror -Wno-strict-aliasing"
+   WARN_CFLAGS="-Wall -W -Wshadow -Wstrict-prototypes -Wpointer-arith -Wcast-align -Wwrite-strings -Wno-strict-aliasing"
 fi
 AC_SUBST(WARN_CFLAGS)
 
