--- libiscsi-1.9.0/lib/socket.c~	2013-06-14 02:37:24.240472976 +0200
+++ libiscsi-1.9.0/lib/socket.c	2013-06-14 02:37:26.925516880 +0200
@@ -711,11 +711,14 @@ iscsi_service(struct iscsi_context *iscs
 			return iscsi_service_reconnect_if_loggedin(iscsi);
 		}
 
-		struct sockaddr_in local;
-		socklen_t local_l = sizeof(local);
-		if (getsockname(iscsi->fd, (struct sockaddr *) &local, &local_l) == 0) {
-			ISCSI_LOG(iscsi, 2, "connection established (%s:%u -> %s)", inet_ntoa(local.sin_addr),
-						(unsigned)ntohs(local.sin_port),iscsi->connected_portal);
+		union {
+			struct sockaddr_in local;
+			struct sockaddr sa;
+		} addr;
+		socklen_t local_l = sizeof(addr);
+		if (getsockname(iscsi->fd, &addr.sa, &local_l) == 0) {
+			ISCSI_LOG(iscsi, 2, "connection established (%s:%u -> %s)", inet_ntoa(addr.local.sin_addr),
+						(unsigned)ntohs(addr.local.sin_port),iscsi->connected_portal);
 		}
 
 		iscsi->is_connected = 1;
